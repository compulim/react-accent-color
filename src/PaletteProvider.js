import PropTypes from 'prop-types';
import React     from 'react';

import createPalette from './createPalette';
import Subject       from './Subject';

// Based on https://docs.microsoft.com/en-us/windows/uwp/style/color#color-palette-building-blocks

export default class Palette extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      palette: new Subject({
        accent : props.accent,
        palette: createPalette(props.accent, props.theme),
        theme  : props.theme
      })
    };
  }

  componentWillReceiveProps(nextProps) {
    if (
      nextProps.accent !== this.props.accent
      || nextProps.theme !== this.props.theme
    ) {
      this.state.palette.next({
        accent : nextProps.accent,
        palette: createPalette(nextProps.accent, nextProps.theme),
        theme  : nextProps.theme
      });
    }
  }

  getChildContext() {
    return this.state;
  }

  render() {
    return this.props.children;
  }
}

Palette.childContextTypes = {
  palette: PropTypes.instanceOf(Subject)
};

Palette.defaultProps = {
  accent: '#0078D7',
  theme : 'light'
};

Palette.propTypes = {
  accent: PropTypes.string,
  theme : PropTypes.oneOf(['dark', 'light'])
};

Palette.defaultAccents = [
  '#FFB900',
  '#FF8C00',
  '#F7630C',
  '#CA5010',
  '#DA3B01',
  '#EF6950',
  '#D13438',
  '#FF4343',
  '#E74856',
  '#E81123',
  '#EA005E',
  '#C30052',
  '#E3008C',
  '#BF0077',
  '#C239B3',
  '#9A0089',
  '#0078D7',
  '#0063B1',
  '#8E8CD8',
  '#6B69D6',
  '#8764B8',
  '#744DA9',
  '#B146C2',
  '#881798',
  '#0099BC',
  '#2D7D9A',
  '#00B7C3',
  '#038387',
  '#00B294',
  '#018574',
  '#00CC6A',
  '#10893E',
  '#7A7574',
  '#5D5A58',
  '#68768A',
  '#515C6B',
  '#567C73',
  '#486860',
  '#498205',
  '#107C10',
  '#767676',
  '#4C4A48',
  '#69797E',
  '#4A5459',
  '#647C64',
  '#525E54',
  '#847574',
  '#7E735F'
].reduce((defaultAccents, color) => {
  defaultAccents[color] = color;

  return defaultAccents;
}, {})
