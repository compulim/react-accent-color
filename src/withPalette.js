import PropTypes     from 'prop-types';
import React         from 'react';

import arrayEqual      from './arrayEqual';
import createPalette   from './createPalette';
import mapEqual        from './mapEqual';
import memoize         from './memoize';
import PaletteProvider from './PaletteProvider';

const DEFAULT_ACCENT = '#0078D7';
const DEFAULT_THEME  = 'light';

function getAccent(props, subject) {
  const subjectValue = subject && subject.getValue();

  return props.accent || (subjectValue && subjectValue.accent) || DEFAULT_ACCENT;
}

function getTheme(props, subject) {
  const subjectValue = subject && subject.getValue();

  return props.theme || (subjectValue && subjectValue.theme) || DEFAULT_THEME;
}

function propsEqual(x, y) {
  const keys = Object.keys(x);

  return (
    arrayEqual(keys, Object.keys(y))
    && keys.every(key, x[key] === y[key])
  );
}

export default function withPalette(propsFactories) {
  return WrappedComponent => {
    const WithPalette = class extends React.Component {
      constructor(props, context) {
        super(props, context);

        const subject = context.palette && context.palette.getValue();

        this.createPalette = memoize(createPalette);
        this.state = this.createState(props, subject, true);
      }

      componentWillMount() {
        this.subscribePalette(this.context);
      }

      componentWillReceiveProps(nextProps, nextContext) {
        let shouldUpdate;

        if (nextContext.palette !== this.context.palette) {
          this.subscribePalette(nextContext);
          shouldUpdate = 1;
        }

        // React may fire componentWillReceiveProps more than it should
        // https://github.com/facebook/react/issues/3610
        if (!mapEqual(nextProps, this.props)) {
          shouldUpdate = 1;
        }

        shouldUpdate && this.setState(() => this.createState(nextProps, nextContext.palette && nextContext.palette.getValue(), true));
      }

      subscribePalette(context) {
        this.unsubscribe && this.unsubscribe();

        if (context.palette) {
          this.unsubscribe = context.palette.subscribe(palette => {
            (!this.props.accent || !this.props.theme) && this.setState(() => this.createState(this.props, palette));
          });
        }
      }

      createState(props, subjectValue = {}, propsChanged = false) {
        const nextPalette = this.createPalette(
          props.accent || subjectValue.accent || DEFAULT_ACCENT,
          props.theme || subjectValue.theme || DEFAULT_THEME
        );

        if (propsChanged || nextPalette !== this.state.palette) {
          return {
            cssProps: propsFactories && propsFactories(nextPalette, props),
            palette : nextPalette
          };
        } else {
          return {};
        }
      }

      render() {
        return (
          <WrappedComponent
            { ...this.state.cssProps }
            { ...this.props }
          />
        );
      }
    };

    WithPalette.contextTypes = PaletteProvider.childContextTypes;
    WithPalette.displayName = `withPalette(${ WrappedComponent.displayName || WrappedComponent.name })`;
    WithPalette.propTypes = {
      accent: PropTypes.string,
      theme : PropTypes.oneOf(['dark', 'light'])
    };

    return WithPalette;
  };
}
