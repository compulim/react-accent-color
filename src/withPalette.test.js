import { configure, mount } from 'enzyme';
import Adapter              from 'enzyme-adapter-react-16';
import React                from 'react';
import renderer             from 'react-test-renderer';

import PaletteProvider from './PaletteProvider';
import withPalette     from './withPalette';

configure({ adapter: new Adapter() });

const Dummy = props => <div data-accent={ props.accent } />;

Dummy.displayName = 'Dummy';

it('should wrap component with CSS factories', () => {
  const WrappedDummy = withPalette((palette, props) => ({
    accent             : palette.accent,
    theme              : palette.theme,
    hoistedFourFiveSize: props.fourFiveSix
  }))(Dummy);

  const provider = mount(
    <PaletteProvider accent="#F00" theme="light">
      <WrappedDummy oneTwoThree="123" fourFiveSix="456" />
    </PaletteProvider>
  );

  expect(provider.find('Dummy').props().oneTwoThree).toBe('123');
  expect(provider.find('Dummy').props().accent).toBe('#F00');
  expect(provider.find('Dummy').props().theme).toBe('light');
  expect(provider.find('Dummy').props().hoistedFourFiveSize).toBe('456');
});

it('should refresh with accent color change', () => {
  const paletteFactory = jest.fn();
  const WrappedDummy = withPalette((palette, props) => {
    paletteFactory();

    return {
      myAccent: palette.accent
    };
  })(Dummy);

  const provider = mount(
    <PaletteProvider accent="#F00">
      <WrappedDummy oneTwoThree="123" />
    </PaletteProvider>
  );

  expect(paletteFactory).toHaveBeenCalledTimes(1);

  provider.setProps({ accent: '#0F0' });

  expect(provider.find('Dummy').props().oneTwoThree).toBe('123');
  expect(provider.find('Dummy').props().myAccent).toBe('#0F0');

  // React may call componentWillReceiveProps more than it should, so we can't say it called 2 times exactly
  expect(paletteFactory).toHaveBeenCalledTimes(2);
});

it('should not refresh with no color change', () => {
  const paletteFactory = jest.fn();
  const WrappedDummy = withPalette((palette, props) => {
    paletteFactory();

    return {};
  })(Dummy);

  const provider = mount(
    <PaletteProvider accent="#F00" theme="light">
      <WrappedDummy oneTwoThree="123" />
    </PaletteProvider>
  );

  expect(paletteFactory).toHaveBeenCalledTimes(1);

  provider.setProps({ accent: '#F00', theme: 'light' });

  // We should not call paletteFactory if there is nothing actually changed
  expect(paletteFactory).toHaveBeenCalledTimes(1);
});

it('should override palette with accent props', () => {
  const paletteFactory = jest.fn();
  const WrappedDummy = withPalette((palette, props) => {
    paletteFactory();

    return { myAccent: palette.accent + palette.theme };
  })(Dummy);

  const provider = mount(
    <PaletteProvider accent="#F00" theme="dark">
      <WrappedDummy accent="#999" />
    </PaletteProvider>
  );

  expect(provider.getDOMNode()).toMatchSnapshot();
  expect(provider.find('Dummy').props().myAccent).toBe('#999dark');

  provider.setProps({ accent: '#FFF' });

  // Changing provider.accent should not trigger a new update because we have already overrode the accent color
  expect(paletteFactory).toHaveBeenCalledTimes(1);

  provider.setProps({ dummy: '123' });

  // Changing any unrelated provider.props should not trigger any update
  expect(paletteFactory).toHaveBeenCalledTimes(1);

  provider.setProps({ theme: 'light' });

  expect(paletteFactory).toHaveBeenCalledTimes(2);
  expect(provider.find('Dummy').props().myAccent).toBe('#999light');
});

it('should recalculate palette on props change', () => {
  const paletteFactory = jest.fn();
  const WrappedDummy = withPalette((palette, props) => {
    paletteFactory();

    return { myAccent: palette.accent + (props.dummy || '') };
  })(Dummy);

  const wrappedDummy = mount(<WrappedDummy accent="#999" />);

  expect(paletteFactory).toHaveBeenCalledTimes(1);
  expect(wrappedDummy.find('Dummy').props().myAccent).toBe('#999');

  wrappedDummy.setProps({ dummy: 123 });

  expect(paletteFactory).toHaveBeenCalledTimes(2);
  expect(wrappedDummy.find('Dummy').props().myAccent).toBe('#999123');
});

it('should work without <PaletteProvider>', () => {
  const WrappedDummy = withPalette((palette, props) => ({
    accent             : palette.accent,
    theme              : palette.theme,
    hoistedFourFiveSize: props.fourFiveSix
  }))(Dummy);

  const wrappedDummy = mount(
    <WrappedDummy accent="#F00" theme="dark" oneTwoThree="123" fourFiveSix="456" />
  );

  expect(wrappedDummy.find('Dummy').props().oneTwoThree).toBe('123');
  expect(wrappedDummy.find('Dummy').props().accent).toBe('#F00');
  expect(wrappedDummy.find('Dummy').props().theme).toBe('dark');
  expect(wrappedDummy.find('Dummy').props().hoistedFourFiveSize).toBe('456');
});
