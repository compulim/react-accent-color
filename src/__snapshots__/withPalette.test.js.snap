// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should override palette with accent props 1`] = `
<div
  data-accent="#999"
/>
`;
